File: Problems/LeetCode/C457/A.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef double db;
typedef pair<int,int> PII;
typedef unsigned long long ull;
const int inf = 1000000000;

class Solution {
public:
    struct info {
        string s;
        int id;
        friend bool operator < (const info &a, const info &b) {
            if(a.id != b.id) return a.id < b.id;
            return a.s < b.s;
        }
    };
    vector<string> validateCoupons(vector<string>& a, vector<string>& b, vector<bool>& tag) {
        int n = a.size();
        map<string,int> mp;
        mp["electronics"] = 1;
        mp["grocery"] = 2;
        mp["pharmacy"] = 3;
        mp["restaurant"] = 4;
        vector<info> t;
        for(int i = 0; i < n; i++) {
            if(a[i].empty() || !tag[i] || !mp[b[i]])  continue;
            bool ok = true;
            for(char c : a[i]) {
                if(!isalnum(c) && c != '_') {
                    ok = false;
                    break;
                }
            }
            if(!ok) continue;
            t.emplace_back(a[i], mp[b[i]]);
        }
        sort(t.begin(), t.end());
        vector<string> ans;
        for(auto &[s, _] : t) {
            ans.push_back(s);
        }
        return ans;
    }
};

File: Problems/LeetCode/C457/A.py
mp = {
    "electronics": 0,
    "grocery": 1,
    "pharmacy": 2,
    "restaurant": 3,
}

class info:
    def __init__(self, s: str, id: int):
        self.s = s
        self.id = id
    def __lt__(self, other):
        if self.id != other.id:
            return self.id < other.id
        return self.s < other.s
class Solution:
    def validateCoupons(self, a: List[str], b: List[str], tag: List[bool]) -> List[str]:
        n = len(a)
        res = []
        for s, t, flag in zip(a, b, tag):
            id = mp.get(t, -1)
            if s and id >= 0 and flag and all(c == '_' or c.isalnum() for c in s):
                res.append(info(s, u))
        res.sort()
        ans = [o.s for o in res]
        return ans
